### PR TITLE
Add ScriptCompDesc

### PR DESCRIPTION
**Pull request by JoneLau**

### CODE REVIEW

**Comment by JoneLau** [comment]
update cocos-creator/engine-3d#227
make inspector list can refer to class ReorderableList and http://va.lent.in/unity-make-your-lists-functional-with-reorderablelist/

File: Assets/Editor/ScriptCompDescEditor.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+using UnityEngine;
+using UnityEditor;
+using UnityEditorInternal;
+using System.IO;
+
+namespace exsdk {
+  [CustomEditor(typeof(ScriptCompDesc))]
+  public class ScriptCompDescEditor : Editor {
+
+    ReorderableList m_List;
```
Comment (jwu) [inline Assets/Editor/ScriptCompDescEditor.cs]: m_List => list

File: Assets/Editor/ScriptCompDescEditor.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    ScriptCompDesc m_Target;
```
Comment (jwu) [inline Assets/Editor/ScriptCompDescEditor.cs]: m_Target => target

File: Assets/Scripts/ScriptCompDesc.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+[System.Serializable]
+[CreateAssetMenu(menuName = "CustomScriptComp/Editor")]
+public class ScriptCompDesc : ScriptableObject {
+  public List<ScriptCompDescProperty> propertyList = new List<ScriptCompDescProperty>();
```
Comment (jwu) [inline Assets/Scripts/ScriptCompDesc.cs]: propertyList => properties

File: Assets/Scripts/ScriptCompDesc.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+}
+[System.Serializable]
+public class ScriptCompDescProperty {
+  public enum TypeMode {
```
Comment (jwu) [inline Assets/Scripts/ScriptCompDesc.cs]: TypeMode => PropType

File: Assets/Scripts/ScriptComponent.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEditor;
+
+public class ScriptComponent : MonoBehaviour {
+
+  ScriptCompDesc desc;
+  public Dictionary<string, object> _properties;
```
Comment (jwu) [inline Assets/Scripts/ScriptComponent.cs]: _properties => properties

File: Assets/Scripts/ScriptComponent.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+  public void initProperties() {
+    if (desc == null) {
```
Comment (jwu) [inline Assets/Scripts/ScriptComponent.cs]: no need for this

File: Assets/Editor/ScriptComponentEditor.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    public override void OnInspectorGUI() {
+      if (properties != null) {
+        foreach (KeyValuePair<string, object> item in show._properties) {
+          properties[item.Key] = item.Value;
+        }
+        foreach (string name in properties.Keys) {
+          EditorGUILayout.BeginHorizontal(new GUILayoutOption[0]);
+          EditorGUILayout.LabelField(name, GUILayout.MaxWidth(50));
+          object value = properties[name];
+          switch (value.GetType().ToString()) {
```
Comment (jwu) [inline Assets/Editor/ScriptComponentEditor.cs]: value.GetType() 即可，不需要 ToString

File: Assets/Editor/ScriptComponentEditor.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+            case "System.String":
+              show._properties[name] = EditorGUILayout.TextField(value.ToString(), GUILayout.MaxWidth(200));
```
Comment (jwu) [inline Assets/Editor/ScriptComponentEditor.cs]: value as String

File: Assets/Editor/ScriptComponentEditor.cs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+              break;
+            case "System.Int32":
+              show._properties[name] = EditorGUILayout.IntField(int.Parse(value.ToString()), GUILayout.MaxWidth(200));
```
Comment (jwu) [inline Assets/Editor/ScriptComponentEditor.cs]: value as Int

File: Assets/Scripts/ScriptComponent.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEditor;
+
+public class ScriptComponent : MonoBehaviour {
+
+  ScriptCompDesc desc;
```
Comment (jwu) [inline Assets/Scripts/ScriptComponent.cs]: public ScriptCompDesc

File: Assets/Editor/ScriptComponentEditor.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace exsdk {
+  [CustomEditor(typeof(ScriptComponent))]
+  public class ScriptComponentEditor : Editor {
+    ScriptComponent show;
+    private Dictionary<string, object> properties;
+    private void OnEnable() {
+      show = (ScriptComponent)target;
+      properties = new Dictionary<string, object>();
+      show.initProperties();
```
Comment (jwu) [inline Assets/Editor/ScriptComponentEditor.cs]: 不应该初始化 properties，而是检查 desc 是否有改动，如果有改动，则 resetProperties

File: Assets/Editor/Exporter.Scene.cs [hunk 12]
```diff
@@ -340,6 +340,22 @@ public partial class Exporter {
         result.Add(comp);
       }
 
+      // script-component
+      ScriptComponent scriptComp = _go.GetComponent<ScriptComponent>();
+      if (scriptComp) {
+        JSON_Component comp = new JSON_Component();
+        comp.type = "Script";
```
Comment (jwu) [inline Assets/Editor/Exporter.Scene.cs]: comp.type = _go.desc.name;

**Pull request merged**